src/pybind/mgr/ansible_orchestrator/ansible_runner_svc.py

```diff
@@ -0,0 +1,278 @@
+
+    def get_result(self, event_filter=""):
+        """Get the data of the events filtered by a task pattern and
+        a event filter
+
+        @returns: the events that matches with the patterns provided
+        """
+
+        if not self.result_task_pattern or not self.play_uuid:
+            result_events = {}
+
+        response = self.rest_client.http_get(PLAYBOOK_EVENTS % self.play_uuid)
+
+        if not response:
+            result_events = {}
+        else:
+            events = json.loads(response.text)["data"]["events"]
+            result_events = {event:data for event,data in events.iteritems()
+                            if "task" in data and
+                            re.match(self.result_task_pattern, data["task"])}
+            if event_filter:
+                result_events = {event:data for event,data in result_events.iteritems()
+                                if re.match(event_filter, data['event'])}
+
+        self.log.info("Requested playbook result is: %s", json.dumps(result_events))
+        return result_events
+
+class Client(object):
+    """An utility object that allows to connect with the Ansible runner service
+    and execute easily playbooks
+    """
+
+    def __init__(self, server, port, user, password, certificate, logger):
+        """Get the Ansible runner service url and user credentials to be used in
+        all the requests.
+
+        @param ar_service_url: The URL of the Ansible runner service
+        @param user: A tuple with (user_name, password), the user authorized to
+                    connect with the Ansible Runner service
+        """
+        self.server = server
+        self.port = port
+        self.user = user
+        self.password = password
+        self.log = logger
+        self.auth = (self.user, self.password)
+
+        # Used to verify or not https server identity
+        if not certificate:
+            self.certificate = False
+        else:
+            self.certificate = certificate
+
+        # Once authenticated this token will be used in all the requests
+        self.token = ""
+
+        self.server_url = "https://{0}:{1}".format(self.server, self.port)
+
+        # Log in the server and get a token
+        self.login()
+
+    def login(self):
+        """ Login with user credentials to obtain a valid token
+        """
+
+        response = None
+
+        try:
+            the_url = "%s/%s" % (self.server_url, LOGIN_URL)
+            r = requests.get(the_url,
+                             auth = self.auth,
+                             verify = self.certificate)
+
+            if r.status_code != requests.codes.ok:
+                self.log.error("login error <<%s>> (%s):%s",
+                                the_url, r.status_code, r.text)
+            else:
+                self.log.info("login succesful <<%s>> (%s):%s",
+                               the_url, r.status_code, r.text)
+
+            response = r
+
+        except Exception as ex:
```

Comment:
Catching `Exception` is a bit broad. Do you want to catch more than`requests.exceptions.RequestException` here?

Reply:
No special action will be taken in case of any kind of error here , so in my opinion differenciate the errors is not adding too much value.
In any case, following your advice, I changed the log method to ""exception" in order to have more information about "the context" of the error

Reply:
The problem with just logging exceptions is that the ceph-mgr log is not visible to users unless they go and poke around on the node where it's running.  To actually make an error visible to users, it's better to let the exception surface.

In other words, if there's no special action to take in the case of the exception, then don't catch it.  In the case of login(), it makes sense to have the caller catch exceptions, rather than to catch them inside and then have the caller check is_operable -- that way the caller can see the actual exception, and surface it to the user (e.g. via a health check) if they choose to.

Reply:
I have checked the behavior of the module with the login implementation with/without error management. 
And In this case (login), as John is pointing, it seems sensible to move the error management to the caller. 
(not too much difference, but the error message/stack trace is clear because the explanation appears first).

In the case of the rest of the http methods, i think that is better to leave the error management in the method, if i remove it, then probably a generic error management will be implemented at the caller level to avoid repeating the error management in each http call.


Details:

I stopped the Ansible Runner Service to check the error:

When i try to login again: ( error management in login method)
```
2018-10-24 14:38:40.301 7fa08c337700  0 mgr[ansible_orchestrator] Ansible runner service - Unexpected error
Traceback (most recent call last):
  File "/usr/lib64/ceph/mgr/ansible_orchestrator/ansible_runner_svc.py", line 171, in login
    verify = self.certificate)
  File "/usr/lib/python2.7/site-packages/requests/api.py", line 68, in get
    return request('get', url, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/api.py", line 50, in request
    response = session.request(method=method, url=url, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 464, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 576, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/adapters.py", line 415, in send
    raise ConnectionError(err, request=request)
ConnectionError: ('Connection aborted.', error(111, 'Connection refused'))
2018-10-24 14:38:40.301 7fa08c337700  0 mgr[ansible_orchestrator] Ansible Runner Service not available. Check external server status or connection options. If configuration options changed try to disable/enable the module.
```

When i try to login again: (login method without error management, it is implemented in the caller.) (This is the current version)

```
2018-10-24 15:05:31.908 7f1056d41700  0 mgr[ansible_orchestrator] Ansible Runner Service not available. Check external server status or connection options. If configuration options changed try to disable/enable the module.
Traceback (most recent call last):
  File "/usr/lib64/ceph/mgr/ansible_orchestrator/module.py", line 250, in serve
    logger = self.log)
  File "/usr/lib64/ceph/mgr/ansible_orchestrator/ansible_runner_svc.py", line 159, in __init__
    self.login()
  File "/usr/lib64/ceph/mgr/ansible_orchestrator/ansible_runner_svc.py", line 171, in login
    verify = self.certificate)
  File "/usr/lib/python2.7/site-packages/requests/api.py", line 68, in get
    return request('get', url, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/api.py", line 50, in request
    response = session.request(method=method, url=url, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 464, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python2.7/site-packages/requests/sessions.py", line 576, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python2.7/site-packages/requests/adapters.py", line 415, in send
    raise ConnectionError(err, request=request)
ConnectionError: ('Connection aborted.', error(111, 'Connection refused'))
```